### PR TITLE
[SDK-426] Implement a minimally viable frontend for FML

### PR DESCRIPTION
**Pull request by travis79**
rebase cruft

### CODE REVIEW

File: components/support/nimbus-fml/src/fixtures/intermediate_representation.rs [hunk 1]
```diff
@@ -95,14 +95,38 @@ mod dump_to_file {
         Ok(())
     }
 
+    fn write_yaml(fm: &FeatureManifest, nm: &str) -> Result<()> {
+        let root = std::env::var("CARGO_MANIFEST_DIR")
+            .expect("Missing $CARGO_MANIFEST_DIR, cannot write fixtures files");
+        let fixtures_dir = "fixtures";
+        let path: PathBuf = [&root, fixtures_dir, nm].iter().collect();
```
Comment (tarikeshaq) [inline components/support/nimbus-fml/src/fixtures/intermediate_representation.rs:102]: super nit: we can isolate a `get_fixture_path` function and use it in both `write_*` functions (or use generics, but with only two use cases, defs not worth it)

**Review by tarikeshaq — APPROVED** [review]
Looks good to me!

File: components/support/nimbus-fml/src/parser.rs [hunk 2]
```diff
@@ -2,12 +2,237 @@
...
+            "Enum" => Ok(TypeRef::Enum(type_name.unwrap())),
+            "Object" => Ok(TypeRef::Object(type_name.unwrap())),
+            "List" => Ok(TypeRef::List(Box::new(TypeRef::try_from(
+                type_name.unwrap(),
+            )?))),
+            "Option" => Ok(TypeRef::Option(Box::new(TypeRef::try_from(
+                type_name.unwrap(),
+            )?))),
+            "Map" => {
+                // Maps take a little extra massaging to get the key and value types
```
Comment (jhugman) [inline components/support/nimbus-fml/src/parser.rs:49]: Yeah, this looks narlier than it needs to be :(

Let's get rid of `StringMap` and rename `EnumMap` to `Map`. What was I thinking.

File: components/support/nimbus-fml/fixtures/nimbus_features.yaml [hunk 3]
```diff
@@ -0,0 +1,34 @@
+---
```
Comment (jhugman) [inline components/support/nimbus-fml/fixtures/nimbus_features.yaml:1]: Could we translate the IR into YAML and put it in a directory in `fixtures`, and then generate the code again and run the same tests that are running off of the IR?

I think `workflows/mod.rs` should be the place to do the running of the tests.

**Review by jhugman — APPROVED** [review]
Nice work.

I've added feedback below, but nothing that should hold you up another review cycle.

File: components/support/nimbus-fml/fixtures/nimbus_features.yaml [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+            "jump-back-in": "false"
+            "recently-saved": "false"
+          ]
+    default: |
+      (
+        "sections-enabled": [
+          "top-sites": "true"
+          "jump-back-in": "false"
+          "recently-saved": "false"
+        ]
```
[the remark this answers is not in the document]
Comment (travis79) [reply components/support/nimbus-fml/fixtures/nimbus_features.yaml:33]: We can, I just hadn't determined exactly what I wanted the defaults to look like yet, so I opted for a JSON blob. You are right, the square brackets probably should be curly braces...

File: components/support/nimbus-fml/src/parser.rs [hunk 5]
```diff
@@ -2,12 +2,209 @@
...
+#[derive(Debug, Deserialize, Serialize)]
+pub(crate) struct FeatureBody {
+    description: String,
+    variables: HashMap<String, FeatureVariableBody>,
+    default: String
+}
+#[derive(Debug, Deserialize, Serialize)]
+pub(crate) struct ManifestFrontEnd {
+    enums: HashMap<String, EnumBody>,
+    features: HashMap<String, FeatureBody>
```
[the remark this answers is not in the document]
Comment (travis79) [reply components/support/nimbus-fml/src/parser.rs]: Coming in the next commit!

**Comment by travis79** [comment]
I need to file an issue to track adding the "Generate IR Manifest for channel" functionality, but this is ready for another look and hopefully landing.

**Review by jhugman — APPROVED** [review]
I have left some comments which should be addressed, but not enough to hold up for another review cycle. Nice work.

**Pull request merged**